src/onefetch/repo.rs

```diff
@@ -0,0 +1,168 @@
+use crate::onefetch::{commit_info::CommitInfo, error::*};
+use git2::{BranchType, Repository, RepositoryOpenFlags, Status, StatusOptions, StatusShow};
+use regex::Regex;
+use std::path::Path;
+
+pub struct Repo {
+    repo: Repository,
+}
+
+impl Repo {
+    pub fn new(repo_path: &str) -> Result<Self> {
+        let repo = Repository::discover(repo_path)?;
+        Ok(Self { repo })
+    }
+    pub fn get_work_dir(&self) -> Result<String> {
+        if let Some(workdir) = self.work_dir()?.to_str() {
+            Ok(workdir.to_string())
+        } else {
+            Err("invalid workdir".into())
+        }
+    }
+
+    pub fn get_number_of_tags(&self) -> Result<usize> {
+        Ok(self.repo.tag_names(None)?.len())
+    }
+
+    pub fn get_number_of_branches(&self) -> Result<usize> {
+        Ok(self.repo.branches(Some(BranchType::Remote))?.count() - 1)
+    }
+
+    pub fn get_git_username(&self) -> Result<String> {
+        let config = self.repo.config()?;
+        let username = config.get_entry("user.name")?;
+        Ok(username.value().unwrap_or("unknown").to_string())
+    }
+
+    pub fn get_version(&self) -> Result<String> {
+        let mut version_name = String::new();
+        let mut most_recent: i64 = 0;
+
+        self.repo.tag_foreach(|id, name| {
+            if let Ok(name) = String::from_utf8(name[10..name.len()].into()) {
+                let mut current_time: i64 = 0;
+                if let Ok(tag) = self.repo.find_tag(id) {
+                    if let Ok(c) = self.repo.find_commit(tag.target_id()) {
+                        current_time = c.time().seconds();
+                    }
+                } else if let Ok(c) = self.repo.find_commit(id) {
+                    current_time = c.time().seconds();
+                }
+                if current_time > most_recent {
+                    most_recent = current_time;
+                    version_name = name;
+                }
+
+                return true;
+            }
+            false
+        })?;
+
+        Ok(version_name)
+    }
+
+    pub fn get_pending_changes(&self) -> Result<String> {
+        let statuses = self.repo.statuses(Some(
+            StatusOptions::default()
+                .show(StatusShow::Workdir)
+                .update_index(true)
+                .include_untracked(true)
+                .renames_head_to_index(true)
+                .recurse_untracked_dirs(true),
+        ))?;
+
+        let mut deleted: usize = 0;
+        let mut added: usize = 0;
+        let mut modified: usize = 0;
+
+        for e in statuses.iter() {
+            let s: Status = e.status();
+            if s.is_index_new() || s.is_wt_new() {
+                added += 1;
+            } else if s.is_index_deleted() || s.is_wt_deleted() {
+                deleted += 1;
+            } else {
+                modified += 1;
+            }
+        }
```

Comment:
This isn't necessarily an *improvement*, just a different way to do the same thing. Totally up to your preference for which gets implemented :slightly_smiling_face:
```suggestion
        let (added, deleted, modified) = statuses.iter().fold((0, 0, 0), |(added, deleted, modified), e| {
            let s: Status = e.status();
            if s.is_index_new() || s.is_wt_new() {
                (added + 1, deleted, modified)
            } else if s.is_index_deleted() || s.is_wt_deleted() {
                (added, deleted + 1, modified)
            } else {
                (added, deleted, modified + 1)
            }
        });
```

Reply:
I love it :heart_eyes_cat: